File: src/main/java/com/vc/medium/L560.scala
package com.vc.medium

object L560 {
  def subarraySum(arr: Array[Int], k: Int): Int = {
    var res = 0
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, Int]()
    map.put(0, 1)
    var sum = 0
    arr.indices.foreach(i => {
      sum += arr(i)
      if(map.contains(sum - k)) {
        res += map(sum - k)
      }
      map.put(sum, map.getOrElse(sum, 0) + 1)
    })
    res
  }
}

File: src/main/java/com/vc/medium/L560.java
package com.vc.medium;

import java.util.HashMap;

class L560 {
    public int subarraySum(int[] arr, int k) {
        if(arr == null || arr.length == 0) return 0;

        HashMap<Integer, Integer> prefix = new HashMap<>();
        prefix.put(0, 1);

        int runningSum = 0, res = 0;
        for(int i = 0; i < arr.length; i++) {
            runningSum += arr[i];
            if(prefix.containsKey(runningSum - k)) {
                res += prefix.get(runningSum - k);
            }
            prefix.put(runningSum, prefix.getOrDefault(runningSum, 0) + 1);
        }
        return res;
    }
}
